### PR TITLE
Ports - Kate & Margaret 

### PR DESCRIPTION
**Pull request by msfinnan**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | We spent a good deal of time on the first day just reading the code and drawing a chart of what we thought was going on. That visual reference was extremely valuable and we were able to get started on building out the code within about an hour. 
What _inheritance_ relations exist between classes? | The CSV class is the parent class to Driver, Trip, and Passenger classes. They inherit data from the CSV class.
What _composition_ relations exist between classes? | Driver, Trip, and Passenger all sit on the same level of hierarchy. An example of related classes is how the request_trip method creates an instance of Trip from the TripDispatcher class because there is a relationship between the two.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | We decided to make a helper method to find an available driver after we recognized that putting that functionality inside of request_trip gave the method too many responsibilities. 
Give an example of a _template method_ that you implemented for this assignment | We used the self.load_all method from CsvRecord and it acted as a template because depending on which class it was called upon, it populated with a completely different set of information and used a different build path. 
Give an example of a _nominal test_ that you wrote for this assignment | In almost every class, we created tests that made sure that the initialize method created something of the proper type. For instance, we have a test called "each item in array is a Trip instance" that checks that our passenger.trips array contains valid instances of Trip.
Give an example of an _edge case test_ that you wrote for this assignment | We'll work on adding edge case tests as we have time! 
What is a concept that you gained more clarity on as you worked on this assignment | How methods from different classes can be called on objects outside of that class by using inheritance and reading in files.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
+require "csv"
+require "time"
+
+require_relative "csv_record"
+
+module RideShare
+  FEE = 1.65
+  PERCENT = 0.8
```
Comment (CheezItMan) [inline lib/driver.rb:8]: Good use of constants!

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+      raise ArgumentError, "Driver must have a status of AVAILABLE or UNAVAILABLE" if status != (:AVAILABLE || :UNAVAILABLE) && status.nil?
+    end
+
+    def add_trip(trip)
+      @trips << trip
+    end
+
+    def average_rating
+      @ratings_array = []
```
Comment (CheezItMan) [inline lib/driver.rb:36]: What about incomplete trips?

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        average_rating = @ratings_array.sum / @ratings_array.length
+        return average_rating.to_f
+      else
+        return 0
+      end
+    end
+
+    def total_revenue
+      @revenue = []
+      @trips.each do |trip|
```
Comment (CheezItMan) [inline lib/driver.rb:50]: Two questions:

1.  What if the trip is incomplete?
2.  What if the trip is less than `FEE`?

File: lib/passenger.rb [hunk 4]
```diff
@@ -16,14 +16,30 @@ def add_trip(trip)
       @trips << trip
     end
 
+    def net_expenditures
+      total_cost = 0
+      trips.each do |trip|
```
Comment (CheezItMan) [inline lib/passenger.rb:21]: Again what about incomplete trips?

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -27,14 +36,42 @@ def inspect
               #{passengers.count} passengers>"
     end
 
+    # def find_next_available_driver
+    #   new_driver = @drivers.find { |driver| driver.status == :AVAILABLE }
+    #   return new_driver
+    # end
+
+    # Option for adding argument error for no driver
+    def find_next_available_driver
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:45]: This is a great helper method.  I would make this one `private`, and I would have this method simply find the driver, not change it's status.  That fits the name of the method and it keeps it doing **one** thing.

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -66,5 +83,15 @@
         expect(trip.passenger.id).must_equal 9
       end
     end
+
+    it "net_expenditure method returns correct total" do
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:87]: I would also have a test with no trips, and another with some incomplete trips.

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -121,5 +122,68 @@ def build_test_dispatcher
...
+
+        expect(trip.passenger.trips).must_include trip
+      end
+
+      it "Adds new trip to dispatcher @trips array" do
+        trip = @dispatcher.request_trip(1)
+        expect(@dispatcher.trips).must_include trip
+      end
+
+      it "Raises an ArgumentError if there are no available drivers" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:180]: 👍

File: specs/trip_spec.rb [hunk 8]
```diff
@@ -40,4 +50,13 @@
       end
     end
   end
+  describe "other methods" do
+    it "calculates the length of the ride in seconds" do
```
Comment (CheezItMan) [inline specs/trip_spec.rb:54]: What about incomplete trips?

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and good commit messages
Answer comprehension questions | Check, but `from_csv` is the template method since the CsvRecord class doesn't implement it and the child classes do.  
**Wave 1** |
Appropriate use of Ruby's `Time` | Check
`Trip` has a helper method to calculate duration | Check
`Passenger` methods to calculate total cost and total time | Check
Complete tests for new behavior | Check, but passenger should also have tests for passengers with no trips.
**Wave 2** |
`Driver` inherits from `CsvRecord`, and implements `from_csv` | Check
`Driver`s and `Trip`s are connected by `TripDispatcher` | Check
`Driver` has method to calculate average rating | Check, but no tests for drivers with no trips.
`Driver` has a method to calculate total revenue | Check, ditto with above
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check, but see my inline comments
Trip creation logic is correctly implemented | Check
New `Trip` is connected to the appropriate `Driver` and `Passenger` | Check
Tests for `request_trip` | Check
Wave 1 and 2 methods ignore in-progress trips | **MISSING**
Tests for Wave 1 and 2 methods w/ in-progress trips | **MISSING**
**Overall** | Not bad, you hit most of the learning goals for the project.  See my inline notes, but except for handling in-progress trips, you did very well.
